File: 2019/day_02/solution.exs
program_orig = File.read!("data.txt") |> String.trim() |> String.split(",") |> Enum.map(&String.to_integer/1)

program_orig = List.replace_at(List.replace_at(program_orig, 1, 12), 2, 2)

defmodule Day2 do
  def run_sequence(program, index \\ 0) do
    opcode = Enum.at(program, index)
    case opcode do
      99 -> Enum.at(program, 0)
      1 -> run_sequence(
        List.replace_at(program, Enum.at(program, index+3), Enum.at(program, Enum.at(program, index+1)) + Enum.at(program, Enum.at(program, index+2))),
        index+4
      )
      2 -> run_sequence(
        List.replace_at(program, Enum.at(program, index+3), Enum.at(program, Enum.at(program, index+1)) * Enum.at(program, Enum.at(program, index+2))),
        index+4
      )
    end
  end

  def search_output(program, output) do
    try do
      for i <- 0..99 do
        for j <- 0..99 do
          if run_sequence(List.replace_at(List.replace_at(program, 1, i), 2, j)) == output do
            throw([i, j])
          else
            [i, j]
          end
        end
      end
    catch
      [noun, verb] -> IO.puts 100*noun+verb
    end
  end
end

IO.inspect Day2.run_sequence(program_orig)
IO.inspect Day2.search_output(program_orig, 19690720)

File: 2019/day_02/solution.py
from copy import deepcopy

with open('data.txt', 'r') as program_file:
    program = [int(e) for e in program_file.read().split(',')]

def run_string(string):
    return ",".join([str(r) for r in run_sequence([int(e) for e in string.split(",")])])

def run_array(array):
    return ",".join([str(r) for r in run_sequence(array)])

def run_sequence(program, index=0):
    opcode = program[index]
    if opcode == 99:
        return program
    else:
        value_a = program[program[index+1]]
        value_b = program[program[index+2]]
        if opcode == 1:
            program[program[index+3]] = value_a+value_b
        else:
            program[program[index+3]] = value_a*value_b
        return run_sequence(program, index+4)

assert run_string("1,0,0,0,99") == "2,0,0,0,99"
assert run_string("2,3,0,3,99") == "2,3,0,6,99"
assert run_string("2,4,4,5,99,0") == "2,4,4,5,99,9801"
assert run_string("1,1,1,4,99,5,6,0,99") == "30,1,1,4,2,5,6,0,99"

# Restore 1202
program_part1 = deepcopy(program)
program_part1[1] = 12
program_part1[2] = 2

print("part 1: %d" % run_sequence(program_part1)[0])


def search_output(program, output):
    for noun in range(100):
        for verb in range(100):
            search_program = deepcopy(program)
            search_program[1] = noun
            search_program[2] = verb
            if run_sequence(search_program)[0] == output:
                return noun, verb

noun, verb = search_output(program, 19690720)
print(noun, verb)
print("part 2: %d" % (100*int(noun)+int(verb)))
